### PR TITLE
Add Scaled Dot Product Attention for FP8

### PR DESCRIPTION
**Pull request by wenscarl**
Scaled Dot Product Attention for FP8. @kaixih  @zhangqiaorjc

### CODE REVIEW

**Comment by kaixih** [comment]
@wenscarl is this ready to review? I can help with that. If possible, can you resolve the conflicts and run the pre-commit for the format consistency?

**Comment by wenscarl** [comment]
It's ready for review and conflict fixed. Thanks!

**Review by superbobry — CHANGES_REQUESTED** [review]
It looks like this PR forked jax/jax/_src/cudnn
/fused_attention_stablehlo.py and then modified it for FP8. Does that sound right? If yes, could we instead add support for FP8 to jax/jax/_src/cudnn
/fused_attention_stablehlo.py directly?

**Comment by wenscarl** [comment]
Yes. I will refactor changes to be in fused_attention_stablehlo.py

File: tests/fused_attention_stablehlo_test.py [hunk 1]
```diff
@@ -543,5 +574,170 @@ def test_sdpa_utils(self):
...
+
+    query_quantized = quantize(query, jnp.float8_e4m3fn, jnp.ones((1,)), jnp.float32)
+    key_quantized = quantize(key, jnp.float8_e4m3fn, jnp.ones((1,)), jnp.float32)
+    value_quantized = quantize(value, jnp.float8_e4m3fn, jnp.ones((1,)), jnp.float32)
+    grad_quantized = quantize(grad, jnp.float8_e4m3fn, jnp.ones((1,)), jnp.float32)
+
+    devices = np.array(jax.local_devices()[:4])
+    devices = devices.reshape((2, 2))
+
+    with Mesh(devices, ("dp", "tp")) as mesh:
```
Comment (yashk2810) [inline tests/fused_attention_stablehlo_test.py:630]: Please don't use the context manager.

Comment (wenscarl) [reply tests/fused_attention_stablehlo_test.py:630]: May I know about the background since the original tests use them.

Comment (yashk2810) [reply tests/fused_attention_stablehlo_test.py:630]: original tests shouldn't have and mesh does not really work with jit. So it's better to remove it and in the other tests too.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 2]
```diff
@@ -200,31 +187,80 @@ def create_dot_product_attention_backend_config(batch,
...
+                                                is_bwd):
+  backend_config = create_dot_product_attention_backend_config_base(batch,
+                                                      num_heads,
+                                                      seq_q,
+                                                      seq_kv,
+                                                      dtype,
+                                                      fmha_scale,
+                                                      mask_type,
+                                                      layout,
+                                                      is_bwd)
```
Comment (mattjj) [inline jax/_src/cudnn/fused_attention_stablehlo.py:239]:
```suggestion
def create_dot_product_attention_fp8_backend_config(
    batch, num_heads, seq_q, seq_kv, dtype, fmha_scale, mask_type, layout, is_bwd):
  backend_config = create_dot_product_attention_backend_config_base(
      batch, num_heads, seq_q, seq_kv, dtype, fmha_scale, mask_type, layout, is_bwd)
```

**Review by mattjj — APPROVED** [review]
Thank you!

File: tests/fused_attention_stablehlo_test.py (hunk 1 above)
Comment (mattjj) [reply tests/fused_attention_stablehlo_test.py:630]: IIUC from Yash, we expect things to Just Work after deleting this `with Mesh(...)` context manager.

Comment (mattjj) [reply tests/fused_attention_stablehlo_test.py:630]: IIUC from Yash, we expect things to Just Work after deleting this `with Mesh(...)` context manager.

**Comment by wenscarl** [comment]
gently ping @mattjj .

**Comment by superbobry** [comment]
Hey @wenscarl, sorry for the silence. I'm a bit swamped this week, but I will try to look through the PR over the weekend or first thing next week.

**Comment by wenscarl** [comment]
> Hey @wenscarl, sorry for the silence. I'm a bit swamped this week, but I will try to look through the PR over the weekend or first thing next week.
Gently ping @superbobry . Thanks!

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 3]
```diff
@@ -1018,56 +1667,92 @@ def dot_product_attention(query: Array,
...
+      raise ValueError(f"Require sliding_window_length > 0, got {sliding_window_length}")
+    if bias is not None:
+      # reshape bias to have 4D shape
+      bias = bias.reshape((1,) * (4 - len(bias.shape)) + bias.shape)
+
+      if mask is not None:
+        if mask.dtype == jnp.bool:
+          large_negative_number = get_large_negative_number(query.dtype)
+          mask = jnp.where(mask, jnp.asarray(0, query.dtype), large_negative_number)
+        # reshape mask to have 4D shape
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:1728]: As above.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 4]
```diff
@@ -1018,56 +1667,92 @@ def dot_product_attention(query: Array,
...
+        mask = mask.reshape((1,) * (4 - len(mask.shape)) + mask.shape)  # type: ignore[union-attr]
+
+      # combine bias and mask
+      if bias is None:
+        bias = mask
+      else:
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:1734]: elif?

File: tests/fused_attention_stablehlo_test.py [hunk 5]
```diff
@@ -543,5 +593,118 @@ def test_sdpa_utils(self):
...
+@jtu.with_config(jax_numpy_dtype_promotion='standard')
+class DotProductAttentionF8Test(jtu.JaxTestCase):
+  def setUp(self):
+    super().setUp()
+    try:
+      cudnn_version = check_cudnn_version()
+    except RuntimeError as e:
+      self.skipTest(str(e))
+      return
+    if cudnn_version < 9010:
```
Comment (superbobry) [inline tests/fused_attention_stablehlo_test.py:606]: This check is probably not necessary (see comment in the other file).

**Review by superbobry — APPROVED** [review]
Please consider sending smaller PRs for any future changes. This is too big to review (well) in one go.

**Comment by superbobry** [comment]
Thanks, can you squash the commits please?

**Comment by wenscarl** [comment]
> Thanks, can you squash the commits please?

Done.

**Review by superbobry — APPROVED** [review]

**Pull request merged**